### ISSUE TITLE
Photo editor

### ISSUE THREAD

**Issue by Donkey-Doug** [deckerst/aves#17]
Add a basic photo editor with support for:

- Cropping (free form, 1:1, 16:9)
- Rotating (I recomment just one button that rotates 90 degrees. Can be pressed multiple times to rotate 180 or 270 degrees)
- Annotating (drawing lines of varying thickness and color, writing text)
- Black-and-White filter to remove colors

**Comment by deckerst** [comment]
Planned but not in the short term, as some other core features and
performance issues must be tackle first.

For rotating and mirroring, you can already do so with the image viewer
menu. There's a line with 3 icon buttons: rotate CCW, rotate CW, and
horizontal flip. They do not modify the image itself, they modify the
metadata, so it's lossless.

**Comment by Subbarao6338** [comment]
Also add lighten images

**Comment by RuthlessRuler** [comment]
Any timeline when the Photo Editing Features will be added?

**Comment by deckerst** [comment]
Nope. I do not have timelines for this personal project. I try to fix bugs as soon as they emerge, but features are just things I'd like to do *someday*.

Is it troublesome to use another app for editing, like Snapseed?

**Comment by bornhustler07** [comment]
Snapseed is close source 🙂

**Comment by bornhustler07** [comment]
If you will consider maybe same day to add photo editor just simple editor kinda like simple gallery app it will be nice👍

**Comment by Darkempire78** [comment]
basic edit is clearly the most important improvements to be made

**Comment by rusty-snake** [comment]
I think Signal's photo editing is the best way to edit photos ATM if you want OSS. Somebody could make an standalone app for it.

**Comment by RuthlessRuler** [comment]
Simple Gallery Pro also has a nice feature rich Photo Editing Tools.

**Comment by mitchwalton** [comment]
Looking forward to a quality photo editor in Aves when he gets to it. However, this app doesn't collect personal info and is powerful photo editor. Give it a try! 

https://apkcombo.com/img-ly-photo-video-editor/com.photoeditorsdk.android.app/

**Comment by Svetov1d** [comment]
built-in editor would be great, add it and i'm switching completely to Aves. I would love to have one app to browse and edit images like simple gallery does

**Comment by arthuragone** [comment]
The best goal of adding a photo editor within Aves would be the possibility not to modify Exif date and position data.
At the moment if you edit your picture with Snapseed as an example, it modify picture date as today and wipe out location data. Same behavior with GrapheneOs internal gallery and editor.

**Comment by opk12** [comment]
What about integrating a lib?
* [PhotoEditor](https://github.com/burhanrashid52/PhotoEditor) does draw, text, filters, stickers, undo / redo.
  It can be embedded as a library and [here](https://github.com/burhanrashid52/PhotoEditor#who-is-using-photoeditor) are the known host apps. Notably, it is embedded in [Screenshot Tile](https://github.com/cvzi/ScreenshotTile) and the F-droid version thereof. So it is F-droid-ready.
  In PhotoEditor's [feature request for cropping](https://github.com/burhanrashid52/PhotoEditor/issues/79), a contributor also suggested uCrop.
* [uCrop](https://github.com/Yalantis/uCrop) does crop, rotate, scale. It can be embedded as a library, however the last commit is very old (is it maintained today?).

**Comment by opk12** [comment]
Also [ImageToolbox](https://github.com/T8RIN/ImageToolbox), is under active development, featureful, already in F-droid.

**Comment by rusty-snake** [comment]
Featureful but lacks in UX.

**Comment by tsutaya0225** [comment]
I agree that the most basic editing functions are necessary, but I don't think anything more than that should be considered now. (I agree with Donkey-Doug who was the first to speak up).
However, I would also like to see crop presets for common photo ratios (3:2, 4:3, 7:5).
I am currently getting by with external apps, so I would be very happy if you could implement this when you can afford it.

**Comment by derei** [comment]
@deckerst I understand there is a lot of work to do ahead, before you may consider adding  new features. However, cropping an image is one of the most used features on mobile devices.
And yes, I'm sure some will say that "annotation", or adding text, stickers, emoji, is what they use most... or color balance, or whatnot. But if we're bein purely objective, the basic ol' crop feature is not only the one present in all photo gallery apps, even those without advanced editing, but is the most used and useful to have at fingertips.

Whilst for advanced editing there are alternatives, I believe that adding a crop feature as a quick modifier, would only benefit the app, and perhaps it could be released before any major update.

Many thanks.

**Comment by Iprobablydontknowwhatimdoing** [comment]
I love snapseed, but it's closed source, and also by google, So they will probably shut it down and they remember it exists.

**Comment by deckerst** [comment]
@Iprobablydontknowwhatimdoing let me know when it's in https://killedbygoogle.com/ :D

**Comment by lordofscripts** [comment]
_Every Android comes with a basic editor_, some lack this or that, others have this and that. 

Personally, I think adding **Yet Another Basic Photo Editor**. would be a waste of valuable resources and bloat ASVES unnecessarily. 

Let AVES excel at what it does.

**Comment by herrbrixter** [comment]
Please add these 2 features:

1. Cropping an image
2. Doodling lines

I used Simple Gallery app before and these are the only editing functions I need.

Thanks!

**Comment by deckerst** [comment]
@herrbrixter I personally only need these too. That would be the first features I work on.

By the way, I've recently discovered that the crop features of Snapseed and Google Photos are not as good as I thought (they downgrade the resolution of very large images). So more incentive for me to implement it and do it right.

**Comment by herrbrixter** [comment]
> @herrbrixter I personally only need these too. That would be the first features I work on.
> 
> By the way, I've recently discovered that the crop features of Snapseed and Google Photos are not as good as I thought (they downgrade the resolution of very large images). So more incentive for me to implement it and do it right.

Thanks @deckerst ! I'd rather have only 1 gallery app in my phone. Now, I have to keep a photo editor app to crop and doodle my pictures.

**Comment by zilexa** [comment]
Have you considered leveraging [libvips](https://github.com/libvips/libvips) for image editing? I am no expert but I believe it's similar but much more advanced and modern compared to ImageMagick. 
It may help you implement editor options swiftly and properly without reinventing the wheel.

**Comment by deckerst** [comment]
Thanks for letting me know.

For the editor in Aves, I need first to decide and implement how it would work (whether it would use layers, undo stack, etc.) and the basic transforms (crop/rotate). Then I would see if/how to integrate with a lib for all the rest.

**Comment by oustenmurh** [comment]
we're still waiting for this feature, aves is only one step to become the best gallery app and that step is this

**Comment by eusi** [comment]
Looking forward to see this basic editing features, especially since Google Photos is bugged for some reason #1421

**Comment by Robert321245** [comment]
I would like to vote on this too. Despite there are several photo editors in Android. It would be handle to have a simple image editor in Aves. The code could be ported from several projects that have an integrated photo editor, like Fossil Gallery and Telegram.

**Comment by opk12** [comment]
Please have proper netiquette and don't spam these "me too" notifications. Instead increment the thumbs up counter on the first post and click `Subscribe` at the right.

**Comment by TheCraftedGamer** [comment]
I read all of this and wonder if there was a way to impliment a 3rd party integration solution. Like if it's hard to make the basic editing software there could be a temporary "open image in ___ app" button.
Even after basic editing software gets implimented, this could still be a useful button to use. 
(Actually the original idea was that I wanted android to have a set default app for editor but thought of an actual temp solution instead lol)

**Comment by rusty-snake** [comment]
How would this be different from the existing "Edit" button?

**Comment by mitchwalton** [comment]
@Donkey-Doug 
@deckerst 

Try "Image Toolbox" as your photo editor of choice with Aves if you haven't heard of it and it has no trackers built into the app. This Image editor is a powerhouse and rich beauty from someone oddly gifted in tech. It pairs with Aves photo edit button flawlessly and opens right up from the app.

https://fdroid.org/packages/ru.tech.imageresizershrinker/

**Comment by SASlobby93** [comment]
> Nope. I do not have timelines for this personal project. I try to fix bugs as soon as they emerge, but features are just things I'd like to do _someday_.
> 
> Is it troublesome to use another app for editing, like Snapseed?


Honestly  i came here after setup of  aves,  then i clicked on edit & found no basic editor in it. i was shocked at first that how come developer forgot to add this feature when the aves have already so many features in it.

then i came here & followed instructions & searched for this & found this & found your comment too. 


then i got another shock, this post raised in 2021 😂
 
& i got my answer.😄
 but to be honest app is too good but its high time to add basic editor  in it.  atleast add cropping & 2-3 filters & annotating. for extra editing we will use another  app but aves should have its basic editor in it. 

Thank you.

**Comment by soymadip** [comment]
> > Nope. I do not have timelines for this personal project. I try to fix bugs as soon as they emerge, but features are just things I'd like to do _someday_.
> > Is it troublesome to use another app for editing, like Snapseed?
> 
> Honestly i came here after setup of aves, then i clicked on edit & found no basic editor in it. i was shocked at first that how come developer forgot to add this feature when the aves have already so many features in it.
> 
> then i came here & followed instructions & searched for this & found this & found your comment too.
> 
> then i got another shock, this post raised in 2021 😂
> 
> & i got my answer.😄 but to be honest app is too good but its high time to add basic editor in it. atleast add cropping & 2-3 filters & annotating. for extra editing we will use another app but aves should have its basic editor in it.
> 
> Thank you.


Yeah true, atleast basic editing with cropping is needed.

Shifting the rotate, flip buttons under edit would be good too

**Comment by deckerst** [comment]
> Shifting the rotate, flip buttons under edit would be good too

Nope, because these operations are lossless, relying on metadata instead of pixel editing.
